File: PythonInterpreter/evaluator.py
'''
The evaluator is where the Lej AST is walked and interpreted.
'''
from typing import Any
from node import Node

###############################################################################
# Assignment Storage
###############################################################################

ASSIGN_MAP: dict[str, Any] = {'unsure_count': 0}

###############################################################################
# Handling val Assignments and Operations
###############################################################################


def build_true() -> list[int]:
    '''
    summary: Build and return the intruitionistically `True` value.
    '''
    return [2]


def build_false() -> list[int]:
    '''
    summary: Build and return the intuitionistically `False` value.
    '''
    return [0]


def build_unsure() -> list[int]:
    '''
    summary: Build and return the intuitionistically `Unsure` value.
        An `Unsure` value houses a unique Boolean truth-table column,
        which is devised based on the number of previous `Unsure` values.
    '''
    unsure_count = ASSIGN_MAP.get('unsure_count', None)
    assert isinstance(unsure_count, int)
    if unsure_count is not None:
        unsure_count += 1
        ASSIGN_MAP['unsure_count'] = unsure_count
        # TODO: Why the hell is this quadrupling instead of doubling?
        u_val_len = 2 << (unsure_count - 1)
        print(f'unsureCount is now {unsure_count},',
              f'and uValLen is now {u_val_len}.')
        u_val: list[int] = [1]
        t = 2
        f = 0
        for _ in range(u_val_len // 2):
            u_val.append(t)
        for _ in range(u_val_len // 2):
            u_val.append(f)
        return u_val
    print('unsure_count is missing.')
    exit(1)
    return []


def double_bools(of_val: list[int]) -> list[int]:
    '''
    summary: Double the length of `of_val` by appending a copy of
        `of_val[1:]` to the end of `of_val`.
    '''
    return of_val + of_val[1:]


def halve_bools(of_val: list[int]) -> list[int]:
    '''
    summary: If the right half of `of_val` is redundant with the left half,
        then return `of_val[:half_size+1]`, where `half_size` is half the
        length of `of_val[1:]`.
        If the right half is not redundant, then return `of_val` unchanged.

        Used for `U`-type vals.
    '''
    if len(of_val) < 4:
        return of_val
    bools_part = of_val[1:]
    half_size = len(bools_part)
    left_half = bools_part[:half_size]
    right_half = bools_part[half_size:]
    for i in range(half_size):
        if left_half[i] != right_half[i]:
            return of_val
    return of_val[:half_size+1]


def gilvenkoize(this_val: list[int]) -> list[int]:
    '''
    summary: If `this_val` only consists of `0`s aside from its first element,
        then return `[0]`, since Gilvenko's theorem applies. Otherwise,
        return `this_val` unchanged.
    '''
    for i in range(1, len(this_val)):
        if this_val[i] > 0:
            return this_val
    return [0]


def build_not(from_val: list[int]) -> list[int]:
    '''
    summary: Return the intuitionistic negation of `from_val`.
    '''
    for i, from_valx in enumerate(from_val):
        from_val[i] = 2 - from_valx
    return from_val


def build_or(val_a: list[int], val_b: list[int]) -> list[int]:
    '''
    summary: Return the intuitionistic disjunction of `val_a` and `val_b`.
    '''
    t = 2
    f = 0
    if val_a[0] == t or val_b[0] == t:
        return [2]
    if val_a[0] == f:
        return val_b
    if val_b[0] == f:
        return val_a
    # Assure that both val_a and val_b are of equal lengths.
    while len(val_a) < len(val_b):
        val_a = double_bools(val_a)
    while len(val_b) < len(val_a):
        val_b = double_bools(val_b)
    for i, (val_ax, val_bx) in enumerate(zip(val_a, val_b)):
        if val_ax < val_bx:
            val_a[i] = val_b[i]
    return gilvenkoize(halve_bools(val_a))


def build_and(val_a: list[int], val_b: list[int]) -> list[int]:
    '''
    summary: Return the intuitionistic conjunction of `val_a` and `val_b`.
    '''
    t = 2
    f = 0
    if val_a[0] == f or val_b[0] == f:
        return [0]
    if val_a[0] == t:
        return val_b
    if val_b[0] == t:
        return val_a
    # Assure that both val_a and val_b are of equal lengths.
    while len(val_a) < len(val_b):
        val_a = double_bools(val_a)
    while len(val_b) < len(val_a):
        val_b = double_bools(val_b)
    for i, (val_ax, val_bx) in enumerate(zip(val_a, val_b)):
        if val_ax > val_bx:
            val_a[i] = val_b[i]
    return gilvenkoize(halve_bools(val_a))


def evaluate_val_expr(val_expr: Node) -> list[int]:
    '''
    summary: Evaluate the VAL-EXPR Nodes and return their appropriate
        intuitionistic valuations.
    '''
    print(f'Evaluating {val_expr}')
    # For already assigned VAL-VARs.
    if val_expr.name == 'VAL-VAR':
        assert isinstance(val_expr.left, Node)
        val_id_key = val_expr.left.literal
        val_id_value: list[int] = ASSIGN_MAP.get(val_id_key, None)
        if val_id_value is None:
            print(val_id_key, 'is not a valid VAL ID.')
            exit(1)
        if not isinstance(val_id_value, list) \
                or not all(isinstance(i, int) and -1 < i < 3
                           for i in val_id_value):
            print(f'{val_id_key} is of the wrong type {val_id_value}.')
            exit(1)
        return val_id_value
    # For all terminals.
    if val_expr.action == 'build_true':
        return build_true()
    if val_expr.action == 'build_false':
        return build_false()
    if val_expr.action == 'build_unsure':
        return build_unsure()
    if val_expr.action == 'LEFT':  # This happens with VAL-ID Nodes.
        assert isinstance(val_expr.left, Node)
        return evaluate_val_expr(val_expr.left)
    # For all nonterminals.
    if val_expr.action == 'build_not':
        assert isinstance(val_expr.left, Node)
        left_side = evaluate_val_expr(val_expr.left)
        return build_not(left_side)
    if val_expr.action == 'build_or':
        assert isinstance(val_expr.left, Node)
        assert isinstance(val_expr.right, Node)
        left_side = evaluate_val_expr(val_expr.left)
        right_side = evaluate_val_expr(val_expr.right)
        return build_or(left_side, right_side)
    if val_expr.action == 'build_and':
        assert isinstance(val_expr.left, Node)
        assert isinstance(val_expr.right, Node)
        left_side = evaluate_val_expr(val_expr.left)
        right_side = evaluate_val_expr(val_expr.right)
        return build_and(left_side, right_side)
    print(f'{val_expr.name} to perform action {val_expr.action}',
          'cannot be correctly evaluated.')
    exit(1)
    return []


def assign_val(this_node: Node) -> None:
    '''
    summary: Assign the value of `this_node.right` to the VAL-VAR
        identified by `this_node.left.left.literal`.
    '''
    assert isinstance(this_node.left, Node)
    assert isinstance(this_node.left.left, Node)
    assert isinstance(this_node.right, Node)
    if this_node.left.name == 'VAL-VAR':
        assign_key: str = this_node.left.left.literal
        assign_value: list[int] = evaluate_val_expr(this_node.right)
        ASSIGN_MAP[assign_key] = evaluate_val_expr(this_node.right)
        print(f'{assign_key} successfully assigned to {assign_value}.')
        return
    print('Invalid type for assign_val.')
    exit(1)

###############################################################################
# Walking the AST
###############################################################################


def walk_tree(this_tree: Node) -> None:
    '''
    summary: Walk the tree and perform the corresponding actions
        on the nodes.
    '''
    if this_tree.action == 'BOTH':
        assert isinstance(this_tree.left, Node)
        assert isinstance(this_tree.right, Node)
        walk_tree(this_tree.left)
        walk_tree(this_tree.right)
        return
    if this_tree.action == 'LEFT':
        assert isinstance(this_tree.left, Node)
        walk_tree(this_tree.left)
        return
    if this_tree.action == 'RIGHT':
        assert isinstance(this_tree.right, Node)
        walk_tree(this_tree.right)
        return
    if this_tree.action == 'assign_val':
        assert isinstance(this_tree, Node)
        assign_val(this_tree)
        return
    print(f'{this_tree.name}\'s {this_tree.action}',
          'has not yet been implemented.')
    exit(1)
    return

File: GoInterpreter/evaluator/evaluator.go
package evaluator

import (
	. "GoInterpreter/node"
	"fmt"
	"os"
)

///////////////////////////////////////////////////////////////////////////////
// Assignment Storage
///////////////////////////////////////////////////////////////////////////////

var AssignMap map[string]interface{} = map[string]interface{}{
	"unsure_count": 0,
}

///////////////////////////////////////////////////////////////////////////////
// Handling val Instantiations and Operations
///////////////////////////////////////////////////////////////////////////////

func buildTrue() []int8 {
	// For terminal "T".
	return []int8{2}
}

func buildFalse() []int8 {
	// For the terminal "F".
	return []int8{0}
}

func buildUnsure() []int8 {
	// For the terminal "U".
	unsureCount, ok := AssignMap["unsure_count"].(int)
	if ok {
		unsureCount++
		AssignMap["unsure_count"] = unsureCount
		// TODO: Why the hell is this quadrupling instead of doubling?
		var uValLen int = 2 << uint(unsureCount-1)
		fmt.Printf("unsureCount is now %d, and uValLen is now %d.\n", unsureCount, uValLen)
		var uVal []int8 = []int8{1}
		var t int8 = 2
		var f int8 = 0
		for i := 0; i < uValLen/2; i++ {
			uVal = append(uVal, t)
		}
		for i := 0; i < uValLen/2; i++ {
			uVal = append(uVal, f)
		}
		return uVal
	}
	print("unsure_count is missing.")
	os.Exit(1)
	return []int8{}
}

func doubleBools(ofVal []int8) []int8 {
	// For "U"-type vals.
	// e.g., [1 2 0] -> [1 2 0 2 0]
	return append(ofVal, ofVal[1:]...)
}

func halveBools(ofVal []int8) []int8 {
	// For "U"-type vals.
	// e.g., [1 2 0 2 0] -> [1 2 0], since half the table is redundant.
	if len(ofVal) == 3 {
		return ofVal
	}
	var boolsPart []int8 = ofVal[1:]
	var halfSize int = len(boolsPart)
	var leftHalf []int8 = boolsPart[:halfSize]
	var rightHalf []int8 = boolsPart[halfSize:]
	for i := 0; i < halfSize; i++ {
		if leftHalf[i] != rightHalf[i] {
			return ofVal
		}
	}
	return ofVal[:halfSize+1]
}

func gilvenkoize(thisVal []int8) []int8 {
	// For "U"-type vals.
	// e.g., [1 0 0 0 0 0 0 0 0] -> [0], since Gilvenko's theorem applies.
	for i := 1; i < len(thisVal); i++ {
		if thisVal[i] > 0 {
			return thisVal
		}
	}
	return []int8{0}
}

func buildNot(fromVal []int8) []int8 {
	for i := 0; i < len(fromVal); i++ {
		fromVal[i] = 2 - fromVal[i]
	}
	return fromVal
}

func buildOr(valA []int8, valB []int8) []int8 {
	// For disjunctions.
	var t int8 = 2
	var f int8 = 0
	if (valA[0] == t) || (valB[0] == t) {
		return []int8{2}
	}
	if valA[0] == f {
		return valB
	}
	if valB[0] == f {
		return valA
	}
	// Assure that both valA and valB are of equal lengths.
	for len(valA) < len(valB) {
		valA = doubleBools(valA)
	}
	for len(valB) < len(valA) {
		valB = doubleBools(valB)
	}
	for i := 0; i < len(valA); i++ {
		if valA[i] < valB[i] {
			valA[i] = valB[i]
		}
	}
	return gilvenkoize(halveBools(valA))
}

func buildAnd(valA []int8, valB []int8) []int8 {
	// For conjunctions.
	var t int8 = 2
	var f int8 = 0
	if (valA[0] == f) || (valB[0] == f) {
		return []int8{0}
	}
	if valA[0] == t {
		return valB
	}
	if valB[0] == t {
		return valA
	}
	// Assure that both valA and valB are of equal lengths.
	for len(valA) < len(valB) {
		valA = doubleBools(halveBools(valA))
	}
	for len(valB) < len(valA) {
		valB = doubleBools(valB)
	}
	for i := 0; i < len(valA); i++ {
		if valA[i] > valB[i] {
			valA[i] = valB[i]
		}
	}
	return gilvenkoize(valA)
}

func evaluateValExpr(valExpr *Node) []int8 {
	// For already assigned VAL-VARs.
	fmt.Printf("Evaluating %v\n", valExpr)
	if valExpr.Name == "VAL-VAR" {
		var valIDKey string = valExpr.Left.Literal
		valIDValue, ok := AssignMap[valIDKey]
		if !ok {
			fmt.Println(valIDKey, " is not a valid VAL ID.")
			os.Exit(1)
		}
		valIDValueChecked, ok := valIDValue.([]int8)
		if !ok {
			fmt.Println(valIDKey, " is of the wrong type ", valIDValueChecked, ".")
			os.Exit(1)
		}
		return valIDValueChecked
	}
	// For all terminals.
	if valExpr.Action == "buildTrue" {
		return buildTrue()
	}
	if valExpr.Action == "buildFalse" {
		return buildFalse()
	}
	if valExpr.Action == "buildUnsure" {
		return buildUnsure()
	}
	if valExpr.Action == "LEFT" {
		return evaluateValExpr(valExpr.Left)
	}
	// For all nonterminals.
	if valExpr.Action == "buildNot" {
		var leftSide []int8 = evaluateValExpr(valExpr.Left)
		return buildNot(leftSide)
	}
	if valExpr.Action == "buildOr" {
		var leftSide []int8 = evaluateValExpr(valExpr.Left)
		var rightSide []int8 = evaluateValExpr(valExpr.Right)
		return buildOr(leftSide, rightSide)
	}
	if valExpr.Action == "buildAnd" {
		var leftSide []int8 = evaluateValExpr(valExpr.Left)
		var rightSide []int8 = evaluateValExpr(valExpr.Right)
		return buildAnd(leftSide, rightSide)
	}
	fmt.Printf("%s to perform action %s cannot be correctly evaluated.\n", valExpr.Name, valExpr.Action)
	os.Exit(1)
	return []int8{}
}

func assignVal(thisNode *Node) {
	if thisNode.Left.Name == "VAL-VAR" {
		var assignKey string = thisNode.Left.Left.Literal
		var assignValue []int8
		assignValue = evaluateValExpr(thisNode.Right)
		AssignMap[assignKey] = evaluateValExpr(thisNode.Right)
		fmt.Printf("%s successfully assigned to %v.\n\n", assignKey, assignValue)
		return
	}
	fmt.Printf("Invalid type for assignVal.\n")
	os.Exit(1)
	return
}

///////////////////////////////////////////////////////////////////////////////
// Walking the AST
///////////////////////////////////////////////////////////////////////////////

func WalkTree(thisTree *Node) {
	if thisTree.Action == "BOTH" {
		WalkTree(thisTree.Left)
		WalkTree(thisTree.Right)
		return
	}
	if thisTree.Action == "LEFT" {
		WalkTree(thisTree.Left)
		return
	}
	if thisTree.Action == "RIGHT" {
		WalkTree(thisTree.Right)
		return
	}
	if thisTree.Action == "assignVal" {
		assignVal(thisTree)
		return
	}
	fmt.Printf("%s's %s has not yet been implemented.\n", thisTree.Name, thisTree.Action)
	os.Exit(1)
	return
}
